### PR TITLE
[Inductor][cond] Allow output-input aliasing in torch.cond branches (re-land D106106380 + dev-nosan test fix)

### PR DESCRIPTION
**Pull request by rmiao**
Summary:
Re-land of D106106380 (reverted by D108642140 - "not approved in OSS, needs more discussion"), folded together with its follow-up dev-nosan test fix D108479020.

Allow output-input aliasing in `torch.cond` branches.

torch.cond branches are mutually exclusive — only one branch executes at runtime. When the true branch returns an operand directly (e.g., identity/skip path), the output aliases the input. This is safe because the non-taken branch never runs, so there is no concurrent access to aliased buffers.

Previously, two checks rejected this pattern:
1. **Dynamo** (`higher_order_ops.py`): `CondHigherOrderVariable.supports_aliasing = False` rejected aliasing during tracing with a graph break.
2. **Functionalization** (`cond.py`): `_check_alias_and_mutation` rejected aliasing during AOT functionalization.

Note: Dynamo already allowed aliasing in inference (`supports_aliasing = not torch.is_grad_enabled()`), but `torch.export` traces with grad enabled even for inference models, so the check was ineffective for the AOTI export path.

Fix:
- Set `supports_aliasing = True` unconditionally on `CondHigherOrderVariable`. Aliasing is safe for cond regardless of grad context because branches are exclusive — even with autograd, only the taken branch's gradient flows.
- Skip the alias check in the functionalization impl (keep mutation check).

Motivation: PrismNet's `_clone_layer_outputs` uses `lo.clone()` to break output-input aliasing required by torch.cond. Removing this constraint saves ~0.046ms per inference (2 DHEN layers at B=2048).

Adds `test_cond_output_input_aliasing` regression test.

Note: D106106380 was reverted for OSS-approval reasons - this re-land needs that discussion resolved (Richard Zou) before it can land.

Fixes T275596551.

Test Plan:
torch.cond aliasing tests (Inductor):
```
buck2 test fbcode//mode/opt -c 'fbcode.nvcc_arch=a100,h100' fbcode//caffe2/test/inductor:control_flow -- --exact 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_device_cpu (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_device_cuda (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_multi_output_device_cpu (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_multi_output_device_cuda (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_both_branches_device_cpu (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_both_branches_device_cuda (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_input_mutation_still_rejected (caffe2.test.inductor.test_control_flow.CondTests)'
```
Test session: https://www.internalfb.com/intern/testinfra/testrun/10414574313882076

PrismNet repro — all 11 tests pass, including `CondOutputInputAlias: without clone` which previously failed with "Replace return input with return input.clone()". Confirms `lo.clone()` workaround in `_clone_layer_outputs` can be removed.
```
buck2 run fbcode//mode/opt fbcode//aps_models/ads/gmp/models/ads_mtml_prism_net_dedicated_model/experimental/benchmarks:repro_aoti_inductor_bugs
```

dev-nosan input-aliasing test (the folded-in D108479020 Sleef fix):
```
buck test fbcode//mode/dev-nosan fbcode//caffe2/test/functorch:test_control_flow -- --exact 'fbcode//caffe2/test/functorch:test_control_flow - test_input_input_alias (caffe2.test.functorch.test_control_flow.TestControlFlowTraced)'
```

Differential Revision: D108709373

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review these changes

File: test/inductor/test_control_flow.py [hunk 1]
```diff
@@ -665,12 +666,106 @@ def false_fn(x, y):
...
+                def true_fn(x):
+                    x.add_(1)
+                    return x
+
+                def false_fn(x):
+                    return x
+
+                return torch.cond(p, true_fn, false_fn, [a])
+
+        with self.assertRaises(Exception):
```
Comment (aorenste) [inline test/inductor/test_control_flow.py]: (Claude Review) [cleanup] `assertRaises(Exception)` is too broad -- it passes for any failure, including an unrelated one (a shape/arg error, an import error, etc.), so it does not actually verify that *mutation* is what gets rejected. Assert the specific type/message this test guards, e.g. `assertRaises(torch._dynamo.exc.UncapturedHigherOrderOpError)` or `assertRaisesRegex(RuntimeError, "might be modifying the input")`.

**Review by zou3519 — CHANGES_REQUESTED** [review]
```
        def true_fn(x):
            return x.clone()

        def false_fn(x):
            view_x = x.view(x.shape)
            return view_x

        def f(x):
            pred = x.shape[0] == 4
            return cond(pred, true_fn, false_fn, [x])
```
we don't support this type of output aliasing because if the user mutates the output of the cond, we need to handle it in functionalization for torch.compile (and the mutation needs to be propagated to the input).

A better solution is we have read-only tensors (and user guarantees us they won't write to the output of the cond), or strict ownership semantics where cond owns the tensors coming in. But those are both a little more complicated to implement.
